wren-core-py/src/extractor.rs

```diff
@@ -0,0 +1,306 @@
+                })
+                .map(|t| t.table().to_string())
+                .collect()
+        })
+}
+
+pub fn extract_manifest(
+    mdl: &Arc<WrenMDL>,
+    used_datasets: Vec<String>,
+) -> Result<PyManifest, CoreError> {
+    let used_models = extract_models(mdl, &used_datasets);
+    let (used_views, models_of_views) = extract_views(mdl, &used_datasets);
+    let used_relationships = extract_relationships(mdl, &used_datasets);
+    Ok(PyManifest {
+        catalog: mdl.catalog().to_string(),
+        schema: mdl.schema().to_string(),
+        models: [used_models, models_of_views].concat(),
+        relationships: used_relationships,
+        metrics: mdl.metrics().to_vec(),
+        views: used_views,
+    })
+}
+
+fn extract_models(mdl: &Arc<WrenMDL>, used_datasets: &[String]) -> Vec<Arc<Model>> {
+    let mut used_set: HashSet<String> = used_datasets.iter().cloned().collect();
+    let mut stack: Vec<String> = used_datasets.to_vec();
+    while let Some(dataset_name) = stack.pop() {
+        if let Some(model) = mdl.get_model(&dataset_name) {
+            model
+                .columns
+                .iter()
+                .filter_map(|col| {
+                    col.relationship
+                        .as_ref()
+                        .and_then(|rel_name| mdl.get_relationship(rel_name))
+                })
+                .flat_map(|rel| rel.models.clone())
+                .filter(|related| used_set.insert(related.clone()))
+                .for_each(|related| stack.push(related));
+        }
+    }
+    mdl.models()
+        .iter()
+        .filter(|model| used_set.contains(model.name()))
+        .cloned()
+        .collect()
+}
+
+fn extract_views(
+    mdl: &Arc<WrenMDL>,
+    used_datasets: &[String],
+) -> (Vec<Arc<View>>, Vec<Arc<Model>>) {
+    let used_set: HashSet<&str> = used_datasets.iter().map(String::as_str).collect();
+    let stack: Vec<&str> = used_datasets.iter().map(String::as_str).collect();
+    let models = stack
+        .iter()
+        .filter_map(|&dataset_name| {
+            mdl.get_view(dataset_name).and_then(|view| {
+                resolve_used_table_names(mdl, view.statement.as_str())
+                    .ok()
+                    .map(|used_tables| extract_models(mdl, &used_tables))
+            })
+        })
+        .flatten()
+        .collect::<Vec<_>>();
+    let views = mdl
+        .views()
+        .iter()
+        .filter(|view| used_set.contains(view.name()))
+        .cloned()
+        .collect();
+
+    (views, models)
+}
+
+fn extract_relationships(
+    mdl: &Arc<WrenMDL>,
+    used_datasets: &[String],
+) -> Vec<Arc<Relationship>> {
+    let mut used_set: HashSet<String> = used_datasets.iter().cloned().collect();
+    let mut stack: Vec<String> = used_datasets.to_vec();
+    while let Some(dataset_name) = stack.pop() {
+        if let Some(relationship) = mdl.get_relationship(&dataset_name) {
```

Comment:
I think it's wrong. `get_relationship` is used to look for a relationship based on the name of the relationship. It's not related to the name of the dataset.
You must search the dataset name in the relationship models to ensure if the relationship is required.

Reply:
By the way, maybe we should add the tests for the relationship.

Reply:
I will fix it. Thanks.